packages/core/src/compiler/Style.ts

```diff
@@ -818,18 +777,36 @@ const matchSelDeclToSubDecl = (
 
   // substitution is only valid if types matched first
   if (isSubtype(subType, selType, domEnv)) {
-    return matchBvar(subAttachedMap, subVar, selVar);
+    return matchBvar(subVar, selVar);
   }
 };
 
 const matchSelDeclToSubDecls = (
   selTypeMap: im.Map<string, Type<A>>,
   subTypeMap: im.Map<string, Type<A>>,
-  subAttachedMap: im.Map<string, LiteralSubExpr<A>>,
   domEnv: DomainEnv,
+  subEnv: SubstanceEnv,
   decl: DeclPattern<A>,
   subProg: CompiledSubProg<A>,
 ): im.List<Subst> => {
+  if (isLiteralType(toDomType(decl.type)) && decl.id.tag === "StyVar") {
+    // special handling for literal types
+
+    // only look at Substance literals with the right type
+    const expectedTag =
```

Comment:
could be brittle in the future but okay for now.

Reply:
the alternative would be to store a mapping from Domain builtins to Substance value types and do a lookup here.

Reply:
Yep when I wrote this code I knew that this can be brittle ...

This also appears in some other places, like

![image](https://github.com/penrose/penrose/assets/16521252/ac6aab55-52bc-4819-a95c-41fa24c31685)


But I think for now this is fine. Will merge ASAP